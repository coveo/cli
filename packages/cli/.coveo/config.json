{"version": 1, "organization": ""}
