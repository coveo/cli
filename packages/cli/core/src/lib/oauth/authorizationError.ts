import dedent from 'ts-dedent';
import type {URL} from 'url';
import {CLIBaseError} from '../errors/CLIBaseError';

export class AuthorizationError extends CLIBaseError {
  public name = 'Authorization Error';
  public constructor(public url: URL) {
    super();

    this.message = [
      url.searchParams.get('error'),
      url.searchParams.get('error_description'),
      url.searchParams.get('error_uri'),
    ]
      .filter((value) => value !== null)
      .join('\n');
  }
}

export class InvalidStateError extends CLIBaseError {
  public name = 'Invalid State Error';
  public constructor(receivedState: string, expectedState: string) {
    super();

    this.message = dedent`Unexpected state in authorization request.
    Expected state: ${expectedState}
    Received state: ${receivedState}`;
  }
}
