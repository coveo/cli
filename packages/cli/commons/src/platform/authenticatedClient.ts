import {setGlobalDispatcher, ProxyAgent, FormData, fetch} from 'undici';
import PlatformClient from '@coveo/platform-client';
import {Config, Configuration} from '../config/config';
import {castEnvironmentToPlatformClient} from './environment';
import globalConfig from '../config/globalConfig';
import {lt} from 'semver';

export class AuthenticatedClient {
  public cfg: Config;
  public constructor() {
    this.cfg = new Config(globalConfig.get().configDir);
  }

  public async isLoggedIn() {
    const {accessToken} = this.cfg.get();
    return accessToken !== undefined;
  }

  public async isExpired() {
    try {
      const c = await this.getClient();
      await c.initialize();
      return false;
    } catch (e) {
      return true;
    }
  }

  public async getClient(overrideConfig?: Partial<Configuration>) {
    const configFromDisk = this.cfg.get();
    const resolvedConfig = {...configFromDisk, ...overrideConfig};
    const globalRequestSettings: Record<string, unknown> = {};
    const proxyServer =
      process.env['https_proxy'] || process.env['HTTPS_PROXY'];
    if (proxyServer) {
      setGlobalDispatcher(new ProxyAgent(proxyServer));
    }
    if (lt(process.version, '18.0.0')) {
      Object.assign(global, {FormData, fetch});
    }
    return new PlatformClient({
      globalRequestSettings,
      environment: castEnvironmentToPlatformClient(resolvedConfig.environment),
      region: resolvedConfig.region,
      organizationId: resolvedConfig.organization,
      accessToken: resolvedConfig.accessToken!,
    });
  }

  public async getAllOrgsUserHasAccessTo() {
    const platformClient = await this.getClient();
    return platformClient.organization.list();
  }

  public async getUserHasAccessToOrg(org: string) {
    const orgs = await this.getAllOrgsUserHasAccessTo();
    return orgs.some((o) => o.id === org);
  }

  public async createImpersonateApiKey(name: string) {
    const platformClient = await this.getClient();
    return await platformClient.apiKey.create({
      displayName: `cli-${name}`,
      description: 'Generated by the Coveo CLI',
      enabled: true,
      privileges: [
        {targetDomain: 'IMPERSONATE', targetId: '*', owner: 'SEARCH_API'},
      ],
    });
  }

  public async getUsername() {
    const {anonymous, organization} = this.cfg.get();
    if (anonymous) {
      return `api-key@${organization}`;
    }
    const authenticatedClient = new AuthenticatedClient();
    const platformClient = await authenticatedClient.getClient();
    await platformClient.initialize();

    return (await platformClient.user.get()).providerUsername;
  }
}

export enum AuthenticationStatus {
  LOGGED_IN,
  EXPIRED,
  LOGGED_OUT,
}

export async function getAuthenticationStatus() {
  const authenticatedClient = new AuthenticatedClient();

  if (!(await authenticatedClient.isLoggedIn())) {
    return AuthenticationStatus.LOGGED_OUT;
  }

  if (await authenticatedClient.isExpired()) {
    return AuthenticationStatus.EXPIRED;
  }

  return AuthenticationStatus.LOGGED_IN;
}
