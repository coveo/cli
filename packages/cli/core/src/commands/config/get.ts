import {CLICommand} from '@coveo/cli-commons/command/cliCommand';
import {Config} from '@coveo/cli-commons/config/config';
import {ConfigRenderer} from '@coveo/cli-commons/config/configRenderer';
import {Trackable} from '@coveo/cli-commons/preconditions/trackable';
import type {Example} from '@oclif/core/lib/interfaces';

export default class Get extends CLICommand {
  public static description = 'Display the current Coveo CLI configuration.';

  public static args = [
    {
      name: 'key',
      description: 'The config key for which to show the value',
      required: false,
    },
  ];

  public static examples: Example[] = [
    {
      command: 'coveo config:get',
      description: 'Get all the configuration values',
    },
    {
      command: 'coveo config:get organization',
      description: 'Get the organization to which you are connected',
    },
    {
      command: 'coveo config:get accessToken',
      description: 'Get the access token given to you by the Coveo Platform',
    },
  ];

  @Trackable()
  public async run() {
    const {args} = await this.parse(Get);
    const cfg = new Config(this.config.configDir);
    const keysToRender = args.key ? [args.key] : undefined;
    ConfigRenderer.render(cfg, keysToRender);
  }
}
