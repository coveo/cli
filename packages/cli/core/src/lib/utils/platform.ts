import {Config} from '@coveo/cli-commons/config/config';

export function getTargetOrg(config: Config, target?: string) {
  if (target) {
    return target;
  }
  const cfg = config.get();
  return cfg.organization;
}
