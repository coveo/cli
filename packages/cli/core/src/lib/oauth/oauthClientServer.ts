import 'fetch-undici-polyfill';
import {createServer, IncomingMessage, Server, ServerResponse} from 'http';
import {URL, URLSearchParams} from 'url';
import {AuthorizationError, InvalidStateError} from './authorizationError';
import {AuthorizationServiceConfiguration, ClientConfig} from './oauthConfig';
interface AccessTokenResponse {
  access_token: string;
}

export class OAuthClientServer {
  public constructor(
    private clientConfig: ClientConfig,
    private authServiceConfig: AuthorizationServiceConfiguration
  ) {}

  public startServer(port: number, hostname: string, expectedState: string) {
    let server: Server;
    const serverPromise = new Promise<{accessToken: string}>(
      (resolve, reject) => {
        const listener = async (req: IncomingMessage, res: ServerResponse) => {
          try {
            const {code, state} = this.parseUrl(req);
            const {tokenEndpoint} = this.authServiceConfig;

            if (!state || !code) {
              return;
            }

            if (state !== expectedState) {
              throw new InvalidStateError(state, expectedState);
            }
            const body = this.getTokenQueryString(code);
            const authRequest = await fetch(tokenEndpoint, {
              method: 'POST',
              body,
              headers: this.requestHeaders,
            });
            const {access_token: accessToken} =
              (await authRequest.json()) as AccessTokenResponse;

            res.end('Close your browser to continue');
            resolve({accessToken});
          } catch (error: unknown) {
            reject(error);
          } finally {
            server.close();
          }
        };

        server = createServer(listener);
        server.listen(port, hostname);
      }
    );

    return serverPromise;
  }

  private get requestHeaders(): Record<string, string> {
    const {client_id} = this.clientConfig;

    return {
      'Content-Type': 'application/x-www-form-urlencoded',
      Authorization: `Basic ${Buffer.from(`${client_id}:${client_id}`).toString(
        'base64'
      )}`,
    };
  }

  private parseUrl(req: IncomingMessage) {
    if (!req.url) {
      return {
        state: null,
        code: null,
      };
    }

    const url = new URL(`http://127.0.0.1${req.url}`);
    const state = url.searchParams.get('state');
    const code = url.searchParams.get('code');
    const error = url.searchParams.get('error');

    if (error) {
      throw new AuthorizationError(url);
    }

    return {state, code};
  }

  private getTokenQueryString(code: string) {
    const {redirect_uri} = this.clientConfig;
    const params = {
      grant_type: 'authorization_code',
      redirect_uri,
      code,
    };

    return new URLSearchParams(params);
  }
}
