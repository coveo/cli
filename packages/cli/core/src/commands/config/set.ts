import {CLICommand} from '@coveo/cli-commons/command/cliCommand';
import {CliUx, Flags} from '@oclif/core';
import {Config} from '@coveo/cli-commons/config/config';
import {AuthenticatedClient} from '@coveo/cli-commons/platform/authenticatedClient';
import {
  IsAuthenticated,
  Preconditions,
} from '@coveo/cli-commons/preconditions/index';
import {Trackable} from '@coveo/cli-commons/preconditions/trackable';
import {InvalidCommandError} from '../../lib/errors/InvalidCommandError';
import {ConfigRenderer} from '@coveo/cli-commons/config/configRenderer';
import type {Example} from '@oclif/core/lib/interfaces';

export default class Set extends CLICommand {
  public static description = 'Modify the current Coveo CLI configuration.';

  public static flags = {
    // TODO CDX-1246
    environment: Flags.string({
      char: 'e',
      hidden: true,
    }),
    // TODO CDX-1246
    region: Flags.string({
      char: 'r',
      hidden: true,
    }),
    organization: Flags.string({
      char: 'o',
      description:
        'The identifier of the organization inside which to perform operations. See <https://docs.coveo.com/en/1562/#organization-id-and-other-information>.',
      helpValue: 'myOrgID',
    }),
  };

  public static examples: Example[] = [
    {
      command: 'coveo config:set --organization myOrgId',
      description: 'connect to a different organization',
    },
  ];

  @Trackable()
  @Preconditions(IsAuthenticated())
  public async run() {
    const {flags} = await this.parse(Set);
    // TODO CDX-1246
    if (flags.environment || flags.region) {
      CliUx.ux.error(
        'To connect to a different region or environment, use the `auth:login` command'
      );
    }
    if (Object.entries(flags).length === 0) {
      throw new InvalidCommandError('Command should contain at least 1 flag');
    }
    const cfg = new Config(this.config.configDir);
    if (flags.organization) {
      await this.verifyOrganization(flags.organization);
      cfg.set('organization', flags.organization);
    }

    ConfigRenderer.render(cfg);
  }

  private async verifyOrganization(org: string) {
    const hasAccess = await new AuthenticatedClient().getUserHasAccessToOrg(
      org
    );
    if (!hasAccess) {
      this.error(
        `You either don't have access to organization ${org}, or it doesn't exist.`
      );
    }
  }
}
