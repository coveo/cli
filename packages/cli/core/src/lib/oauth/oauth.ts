import open from 'open';
import {
  DEFAULT_ENVIRONMENT,
  DEFAULT_REGION,
  PlatformEnvironment,
  platformUrl,
} from '@coveo/cli-commons/lib/platform/environment';
import {Region} from '@coveord/platform-client';
import {randomBytes} from 'crypto';
import {AuthorizationServiceConfiguration, ClientConfig} from './oauthConfig';
import {OAuthClientServer} from './oauthClientServer';
import {URL} from 'url';

export interface OAuthOptions {
  port: number;
  environment: PlatformEnvironment;
  region: Region;
}

export class OAuth {
  private opts: OAuthOptions;

  public constructor(opts?: Partial<OAuthOptions>) {
    const baseOptions: OAuthOptions = {
      port: 32111,
      environment: DEFAULT_ENVIRONMENT,
      region: DEFAULT_REGION,
    };

    this.opts = {
      environment: opts?.environment || baseOptions.environment,
      port: opts?.port || baseOptions.port,
      region: opts?.region || baseOptions.region,
    };
  }

  public async getToken() {
    const state = this.generateState(10);
    const requestHandler = new OAuthClientServer(
      this.clientConfig,
      this.authServiceConfig
    );

    const clientServerPromise = requestHandler.startServer(
      this.opts.port,
      '127.0.0.1',
      state
    );

    this.openLoginPage(state);

    return clientServerPromise;
  }

  private openLoginPage(state: string) {
    const url = this.buildAuthorizationUrl(state);
    open(url).then((browserProcess) => {
      browserProcess.unref();
    });
  }

  private buildAuthorizationUrl(state: string) {
    const {authorizationEndpoint} = this.authServiceConfig;
    const {redirect_uri, client_id, scope} = this.clientConfig;
    const url = new URL(authorizationEndpoint);

    url.searchParams.append('response_type', 'code');
    url.searchParams.append('client_id', client_id);
    url.searchParams.append('redirect_uri', redirect_uri);
    url.searchParams.append('scope', scope);
    url.searchParams.append('state', state);

    return url.href;
  }

  private generateState(size: number) {
    return randomBytes(size).toString('hex');
  }

  private get clientConfig(): ClientConfig {
    return {
      client_id: 'cli',
      redirect_uri: `http://127.0.0.1:${this.opts.port}`,
      scope: 'full',
    };
  }

  private get authServiceConfig(): AuthorizationServiceConfiguration {
    const baseURL = platformUrl({
      environment: this.opts.environment,
      region: this.opts.region,
    });
    return {
      authorizationEndpoint: `${baseURL}/oauth/authorize`,
      revocationEndpoint: `${baseURL}/logout`,
      tokenEndpoint: `${baseURL}/oauth/token`,
    };
  }
}
