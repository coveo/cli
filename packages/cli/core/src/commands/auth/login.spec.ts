jest.mock('../../lib/oauth/oauth');
jest.mock('@coveo/cli-commons/config/config');
jest.mock('@coveo/cli-commons/preconditions/trackable');
jest.mock('@coveo/cli-commons/platform/authenticatedClient');
jest.mock('@coveord/platform-client');
import {Region} from '@coveord/platform-client';
import {test} from '@oclif/test';
import {Config} from '@coveo/cli-commons/config/config';
import {OAuth} from '../../lib/oauth/oauth';
import {AuthenticatedClient} from '@coveo/cli-commons/platform/authenticatedClient';
const mockedOAuth = jest.mocked(OAuth);
const mockedConfig = jest.mocked(Config);
const mockedAuthenticatedClient = jest.mocked(AuthenticatedClient);

describe('auth:login', () => {
  const mockConfigSet = jest.fn();

  const mockConfigGet = jest.fn().mockReturnValue({
    region: 'us',
    organization: 'foo',
    environment: 'prod',
  });

  const mockListOrgs = jest
    .fn()
    .mockReturnValue(Promise.resolve([{id: 'foo'}]));

  const mockGetHasAccessToOrg = jest
    .fn()
    .mockReturnValue(Promise.resolve(true));

  beforeEach(() => {
    mockedAuthenticatedClient.mockImplementation(
      () =>
        ({
          getAllOrgsUserHasAccessTo: mockListOrgs,
          getUserHasAccessToOrg: mockGetHasAccessToOrg,
        } as unknown as AuthenticatedClient)
    );
    mockedOAuth.mockImplementationOnce(
      () =>
        ({
          getToken: () =>
            Promise.resolve({
              accessToken: 'this-is-the-token',
            }),
        } as OAuth)
    );

    mockedConfig.mockImplementation(
      () =>
        ({
          get: mockConfigGet,
          set: mockConfigSet,
        } as unknown as Config)
    );
  });

  afterEach(() => {
    mockConfigSet.mockClear();
  });

  test
    .stdout()
    .stderr()
    .command(['auth:login', '-e', 'foo'])
    .catch(/Expected --environment=foo/)
    .it('reject invalid environment', async () => {});

  test
    .stdout()
    .stderr()
    .command(['auth:login', '-r', 'foo'])
    .catch(/Expected --region=foo/)
    .it('reject invalid region', async () => {});

  ['dev', 'stg', 'prod', 'hipaa'].forEach((environment) => {
    test
      .stdout()
      .stderr()
      .command(['auth:login', '-e', environment, '-o', 'foo'])
      .it(
        `passes the -e=${environment} flag to oauth and configuration`,
        () => {
          expect(mockedOAuth.mock.calls[0][0]?.environment).toBe(environment);
          expect(mockConfigSet).toHaveBeenCalledWith(
            'environment',
            environment
          );
        }
      );
  });

  Object.entries(Region)
    .flatMap((regionEntry) => regionEntry)
    .forEach((region) => {
      test
        .stdout()
        .stderr()
        .command(['auth:login', '-r', region, '-o', 'foo'])
        .it(
          `passes the -e=${region} flag in lowercase to oauth and configuration`,
          () => {
            expect(mockedOAuth.mock.calls[0][0]?.region).toBe(
              region.toLowerCase()
            );
            expect(mockConfigSet).toHaveBeenCalledWith(
              'region',
              region.toLowerCase()
            );
          }
        );
    });

  test
    .stdout()
    .stderr()
    .command(['auth:login', '-e', 'stg', '-o', 'foo'])
    .it('passed the -e=dev flag to oauth as an environment', () => {
      expect(mockedOAuth.mock.calls[0][0]?.environment).toBe('stg');
    });

  describe('retrieves token from oauth service', () => {
    test
      .stdout()
      .stderr()
      .command(['auth:login', '-o', 'foo'])
      .it('save token from oauth service', () => {
        expect(mockConfigSet).toHaveBeenCalledWith(
          'accessToken',
          'this-is-the-token'
        );
      });
    test
      .stdout()
      .stderr()
      .command(['auth:login', '-o', 'foo'])
      .it('set anonymous to false', () => {
        expect(mockConfigSet).toHaveBeenCalledWith('anonymous', false);
      });
  });

  test
    .stdout()
    .stderr()
    .do(() => {
      mockGetHasAccessToOrg.mockReturnValueOnce(Promise.resolve(false));
    })
    .command(['auth:login', '-o', 'this_is_not_a_valid_org'])
    .exit(2)
    .it('fails when organization flag is invalid');

  test
    .do(() => {
      mockListOrgs.mockReturnValueOnce(Promise.resolve([]));
    })
    .stdout()
    .stderr()
    .command(['auth:login'])
    .exit(2)
    .it(
      'fails when no organization flag is passed and the user has access to no organization'
    );

  test
    .stdout()
    .stderr()
    .command(['auth:login', '-o', 'foo'])
    .it('succeed when organization flag is valid', (ctx) => {
      expect(ctx.stdout).toContain('Success');
    });

  test
    .do(() => {
      mockListOrgs.mockReturnValueOnce(
        Promise.resolve([{id: 'the_first_org_available'}])
      );
      mockConfigGet.mockReturnValueOnce({
        region: 'us',
        organization: 'the_first_org_available',
        environment: 'prod',
      });
    })
    .stdout()
    .stderr()
    .command(['auth:login'])
    .it(
      'succeed when no organization flag is passed, and uses the first available org instead',
      (ctx) => {
        expect(ctx.stdout).toContain('Success');
        expect(ctx.stdout).toContain('the_first_org_available');
      }
    );
});
