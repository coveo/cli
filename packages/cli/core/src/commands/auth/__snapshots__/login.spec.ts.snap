// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auth:login succeeds when no organization flag is passed, and uses the first available org instead 1`] = `
"
    Successfully logged in!
    Close your browser to continue.

    You are currently logged in:
    Organization: the_first_org_available
    Region: us
    Environment: prod
    
To change organization, run coveo auth:login --help to see the available options to log in to a different organization, region, or environment.
"
`;

exports[`auth:login succeeds when organization flag is valid 1`] = `
"
    Successfully logged in!
    Close your browser to continue.

    You are currently logged in:
    Organization: foo
    Region: us
    Environment: prod
    
To change organization, either:
  • Run coveo auth:login --help to see the available options to log in to a different organization, region, or environment.
  • Run coveo config:set -o=theOtherOrg to run the next commands against theOtherOrg.
"
`;
